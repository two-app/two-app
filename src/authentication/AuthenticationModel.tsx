export type Tokens = {
  accessToken: string;
  refreshToken: string;
};

export type UserRegistration = {
  uid: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  dob: string;
  acceptedTerms: boolean;
};
