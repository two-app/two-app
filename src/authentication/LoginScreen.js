/**
 * @flow
 */

import React from "react";
import {Button, Text} from "react-native";
import {Wrapper} from "../views/View";

const LoginScreen = ({navigation}) => <Wrapper>
    <Text>Sign In</Text>
    <Button title={"Sign Up"} onPress={() => navigation.navigate('Register')}/>
</Wrapper>;

LoginScreen.navigationOptions = {
    title: 'Sign In',
    header: null
};

export default LoginScreen;