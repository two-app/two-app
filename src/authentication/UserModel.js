// @flow

import jwt_decode from "jwt-decode";

const UnconnectedUser: {
    uid: number,
    connectCode: string
} = {
    uid: null,
    connectCode: null
};

const createUnconnectedUser = (uid: number, connectCode: string): UnconnectedUser => ({uid, connectCode});

const parseUnconnectedUserFromToken = (accessToken: string): (UnconnectedUser) => {
    const decoded = jwt_decode(accessToken);
    const uid = decoded['uid'];
    const connectCode = decoded['connectCode'];

    return createUnconnectedUser(uid, connectCode);
};

export {UnconnectedUser, createUnconnectedUser, parseUnconnectedUserFromToken};