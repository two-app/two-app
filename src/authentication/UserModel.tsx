import jwt_decode from 'jwt-decode';

// @ts-ignore
const decode = (token: string): any => jwt_decode(token);

export type UnconnectedUser = {
    uid: number,
    connectCode: string
}

export const unconnectedUserFromAccessToken = (accessToken: string): UnconnectedUser => ({
    uid: decode(accessToken)['uid'],
    connectCode: decode(accessToken)['connectCode']
});

export type User = {
    uid: number,
    pid: number,
    cid: number
}

export const userFromAccessToken = (accessToken: string): User => ({
    uid: decode(accessToken)['uid'],
    pid: decode(accessToken)['pid'],
    cid: decode(accessToken)['cid']
});