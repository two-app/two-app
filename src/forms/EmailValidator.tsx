/**
 * @param email to verify.
 * @return {boolean} true if the email is valid. False if not, or empty.
 */
function validateEmail(email: string) {
  const re = /^\S+@\S+[.][0-9a-z]+$/;
  return re.test(email);
}

export default {validateEmail};
