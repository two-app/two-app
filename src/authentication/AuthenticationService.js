// @flow
import Gateway from "../http/Gateway";
import {AxiosError, AxiosResponse} from "axios";
import {UserRegistration} from "./register_workflow/UserRegistrationModel";
import {UnconnectedUser} from "./UserModel";
import {Tokens} from "./AuthenticationModel";

class RegisterUserResponse {
    user: UnconnectedUser;
    tokens: Tokens;

    constructor(user: UnconnectedUser, tokens: Tokens) {
        this.user = user;
        this.tokens = tokens;
    }
}


const registerUser = (userRegistration: UserRegistration): Promise<RegisterUserResponse> =>
    Gateway.post("/self", userRegistration)
        .then((r: AxiosResponse) => {
            const accessToken = r.data['accessToken'];
            const refreshToken = r.data['refreshToken'];
            return new RegisterUserResponse({
                user: UnconnectedUser.fromAccessToken(accessToken),
                tokens: new Tokens(accessToken, refreshToken)
            });
        }).catch((e: AxiosError) => {
            throw new Error(e.response.data['message'].toString());
        }
    );

export default {registerUser};
export {RegisterUserResponse};