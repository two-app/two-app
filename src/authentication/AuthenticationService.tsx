import Gateway from '../http/Gateway';
import {AxiosError, AxiosResponse} from 'axios';
import {UserRegistration} from './register_workflow/UserRegistrationModel';
import {UnconnectedUser, unconnectedUserFromAccessToken, User, userFromAccessToken} from './UserModel';
import {Tokens} from './AuthenticationModel';

export type UserResponse = {
    user: UnconnectedUser | User,
    tokens: Tokens
}

const registerUser = (userRegistration: UserRegistration): Promise<UserResponse> => Gateway.post('/self', userRegistration)
    .then((r: AxiosResponse<Tokens>): UserResponse => ({
        user: unconnectedUserFromAccessToken(r.data.accessToken),
        tokens: {accessToken: r.data.accessToken, refreshToken: r.data.refreshToken}
    })).catch((e: AxiosError) => {
        // @ts-ignore
        throw new Error(e.response.data['message'].toString());
    });

const connectToPartner = (connectCode: String): Promise<UserResponse> => Gateway.post(`/partner/${connectCode}`)
    .then((r: AxiosResponse<Tokens>): UserResponse => ({
        user: userFromAccessToken(r.data.accessToken),
        tokens: {accessToken: r.data.accessToken, refreshToken: r.data.refreshToken}
    })).catch((e: AxiosError) => {
        // @ts-ignore
        throw new Error(e.response.data['reason'].toString());
    });

export default {registerUser, connectToPartner};