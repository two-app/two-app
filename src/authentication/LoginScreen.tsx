import React, { useState } from "react";
import { Text, View, TouchableOpacity } from "react-native";
import { ScrollContainer } from "../views/View";
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from '../../Router';
import Colors from "../Colors";
import LogoHeader from "./LogoHeader";
import Input from "../forms/Input";
import UserRegistrationModel from "./register_workflow/UserRegistrationModel";
import SubmitButton from "../forms/SubmitButton";
import AuthenticationService, { LoginCredentials, areCredentialsValid } from "./AuthenticationService";
import { storeTokens } from './store';
import { storeUser, storeUnconnectedUser } from '../user';
import { connect, ConnectedProps } from "react-redux";
import { isUnconnectedUser } from "./UserModel";
import LoadingView from "../views/LoadingView";
import { resetNavigate } from "../navigation/NavigationUtilities";


const mapState = null;
const mapDispatch = { storeUser, storeUnconnectedUser, storeTokens };
const connector = connect(mapState, mapDispatch);
type ConnectorProps = ConnectedProps<typeof connector>;
type LoginScreenProps = ConnectorProps & {
    navigation: StackNavigationProp<RootStackParamList, 'LoginScreen'>;
};

const LoginScreen = ({ navigation, storeUser, storeUnconnectedUser, storeTokens }: LoginScreenProps) => {
    const [loginCredentials, setLoginCredentials] = useState<LoginCredentials>({
        email: '',
        rawPassword: ''
    });
    const [submitted, setSubmitted] = useState(false);
    const [loginError, setLoginError] = useState<string>();

    const login = () => {
        setSubmitted(true)
        AuthenticationService.login(loginCredentials).then(response => {
            storeTokens(response.tokens);
            if (isUnconnectedUser(response.user)) {
                storeUnconnectedUser(response.user);
            } else {
                storeUser(response.user);
            }
            resetNavigate('LoadingScreen', navigation);
        }).catch((e: Error) => {
            setSubmitted(false);
            setLoginError(e.message);
        });
    }

    return (
        <ScrollContainer isLoading={submitted} keyboardShouldPersistTaps='always'>
            <LogoHeader heading="Sign In" />

            <Input attributes={{ placeholder: 'you@email.com', autoCompleteType: 'email', autoCapitalize: 'none' }}
                isValid={UserRegistrationModel.isEmailValid} label={'Email'}
                onChange={email => setLoginCredentials({ ...loginCredentials, email })}
            />

            <Input attributes={{ placeholder: 'Secure Password', autoCompleteType: 'password', secureTextEntry: true }}
                isValid={UserRegistrationModel.isPasswordValid}
                label={'Password'}
                onChange={password => setLoginCredentials({ ...loginCredentials, rawPassword: password })}
            />

            <SubmitButton text="Sign In"
                onSubmit={login}
                disabled={!areCredentialsValid(loginCredentials)}
            />

            {loginError && <Text style={{ color: Colors.DARK_SALMON }} data-testid="error-message">
                {loginError}
            </Text>}

            <View style={{ flexDirection: 'row', marginTop: 10 }}>
                <Text style={{ color: Colors.REGULAR }}>Here to join?</Text>
                <TouchableOpacity style={{ marginLeft: 5 }}
                    onPress={() => resetNavigate('RegisterScreen', navigation)}
                    data-testid={'register-screen-button'}
                >
                    <Text style={{ fontWeight: "bold", color: Colors.DARK }}>Create Account</Text>
                </TouchableOpacity>
            </View>
        </ScrollContainer>
    );
};

export default connector(LoginScreen);
export { LoginScreen }