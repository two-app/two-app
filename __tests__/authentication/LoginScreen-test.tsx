import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { LoginScreen } from '../../src/authentication/LoginScreen';
import { shallow } from 'enzyme';
import { CommonActions } from '@react-navigation/native';
import EmailValidator from '../../src/forms/EmailValidator';
import AuthenticationService, { UserResponse } from '../../src/authentication/AuthenticationService';
import { User, UnconnectedUser } from '../../src/authentication/UserModel';
import { Tokens } from '../../src/authentication/AuthenticationModel';
import { ErrorResponse } from '../../src/http/Response';

describe('LoginScreen', () => {

    test('should maintain snapshot', () => expect(renderer.create(
        <LoginScreen
            navigation={{} as any}
            storeUser={jest.fn()}
            storeUnconnectedUser={jest.fn()}
            storeTokens={jest.fn()}
        />
    ).toJSON()).toMatchSnapshot());

    let tb: LoginScreenTestBed;

    beforeEach(() => tb = new LoginScreenTestBed());

    test('Create Account Button should navigate to RegisterScreen.', () => {
        const btn: any = tb.wrapper.find("TouchableOpacity[data-testid=\'register-screen-button\']").first()

        btn.prop('onPress')();

        expect(tb.dispatchFn).toHaveBeenCalledTimes(1);
        expect(tb.dispatchFn).toHaveBeenCalledWith(
            CommonActions.reset({
                index: 0,
                routes: [{ name: 'RegisterScreen' }]
            })
        );
    });

    describe('Form Validation Tests', () => {
        describe('Email', () => {
            test('delegates to validEmail fn', () => {
                EmailValidator.validateEmail = jest.fn().mockReturnValue(true);
                tb.expectInput('Email', 'admin@two.com');
                expect(EmailValidator.validateEmail).toHaveBeenCalledTimes(1);
                expect(EmailValidator.validateEmail).toHaveBeenCalledWith('admin@two.com');
            });
        });

        describe('Password', () => {
            test('invalid with < 6 characters', () => tb.expectInput('Password', 'pass').toBe(false));
            test('valid with > 6 characters', () => tb.expectInput('Password', 'password').toBe(true));
            test('valid with secure password', () => tb.expectInput('Password', 'P?4Ot2ONz:IJO&%U').toBe(true));
        });

        describe('Valid Form', () => {
            test('should enable the submit button', () => {
                expect(tb.wrapper.find('SubmitButton').prop('disabled')).toBe(true);

                EmailValidator.validateEmail = jest.fn().mockReturnValue(true);
                tb.changeInput('Email', 'admin@two.com');
                tb.changeInput('Password', 'P?4Ot2ONz:IJO&%U');

                expect(tb.wrapper.find('SubmitButton').prop('disabled')).toBe(false);
            });
        });
    });

    describe('Logging In', () => {
        const tokens: Tokens = { accessToken: 'testAccess', refreshToken: 'testRefresh' };

        test('it should call the Authentication Service', () => {
            const user: User = { uid: 1, pid: 2, cid: 3 };
            tb.onLoginReturn({ user, tokens });

            tb.performLogin('user@two.com', 'testPassword');

            expect(AuthenticationService.login).toHaveBeenCalledTimes(1);
        });

        test('it should display a loading overlay', () => {
            tb.onLoginReturn({} as any); // async, result wont be used

            tb.performLogin('user@two.com', 'testPassword');

            expect(tb.wrapper.find("ScrollContainer").prop<boolean>("isLoading")).toBe(true);
        });

        describe('With Valid Response', () => {
            test('it should store the user', done => {
                const user: User = { uid: 1, pid: 2, cid: 3 };
                tb.onLoginReturn({ user, tokens });
                tb.performLogin('user@two.com', 'testPassword');

                setImmediate(() => {
                    expect(tb.storeUserFn).toHaveBeenCalledTimes(1);
                    expect(tb.storeUserFn).toHaveBeenCalledWith(user);
                    done();
                });
            });

            test('it should store the unconnected user', done => {
                const user: UnconnectedUser = { uid: 1, connectCode: 'aBcDeF' };
                tb.onLoginReturn({ user, tokens });
                tb.performLogin('user@two.com', 'testPassword');

                setImmediate(() => {
                    expect(tb.storeUnconnectedUserFn).toHaveBeenCalledTimes(1);
                    expect(tb.storeUnconnectedUserFn).toHaveBeenCalledWith(user);
                    done();
                });
            });

            test('it should store the tokens', done => {
                const user: User = { uid: 1, pid: 2, cid: 3 };
                tb.onLoginReturn({ user, tokens });
                tb.performLogin('user@two.com', 'testPassword');

                setImmediate(() => {
                    expect(tb.storeTokensFn).toHaveBeenCalledTimes(1);
                    expect(tb.storeTokensFn).toHaveBeenCalledWith(tokens);
                    done();
                });
            });

            test('it should navigate to the Loading Screen', done => {
                const user: User = { uid: 1, pid: 2, cid: 3 };
                tb.onLoginReturn({ user, tokens });
                tb.performLogin('user@two.com', 'testPassword');

                setImmediate(() => {
                    expect(tb.dispatchFn).toHaveBeenCalledTimes(1);
                    expect(tb.dispatchFn).toHaveBeenCalledWith(
                        CommonActions.reset({
                            index: 0,
                            routes: [{ name: 'LoadingScreen' }]
                        })
                    );
                    done();
                });
            });
        });

        describe('With Error Response', () => {
            test('it should display the error', done => {
                const error: ErrorResponse = {code: 400, status: '400 Bad Request', reason: 'Some API Error Message'};
                tb.onLoginFail(error);

                tb.performLogin('user@two.com', 'testPassword');

                setImmediate(() => {
                    const errorMessage = tb.wrapper.find("Text[data-testid=\'error-message\']").first();
                    expect(errorMessage.render().text()).toEqual(error.reason);
                    done();
                });
            });
        });
    });

});

class LoginScreenTestBed {

    dispatchFn = jest.fn();
    storeUserFn = jest.fn();
    storeUnconnectedUserFn = jest.fn();
    storeTokensFn = jest.fn();

    wrapper = shallow(
        <LoginScreen
            navigation={{ dispatch: this.dispatchFn } as any}
            storeUser={this.storeUserFn}
            storeUnconnectedUser={this.storeUnconnectedUserFn}
            storeTokens={this.storeTokensFn}
        />
    );

    onLoginReturn = (response: UserResponse) => {
        AuthenticationService.login = jest.fn().mockResolvedValue(response);
    }

    onLoginFail = (error: ErrorResponse) => {
        AuthenticationService.login = jest.fn().mockRejectedValue(error);
    }

    performLogin = (email: string, password: string) => {
        this.changeInput('Email', email);
        this.changeInput('Password', password);
        const onClickFn = this.wrapper.find('SubmitButton').first().prop("onSubmit");
        // @ts-ignore
        onClickFn();
    }

    expectInput = (label: string, value: string) => {
        const input = this.wrapper.find(`Input[label='${label}']`).first();
        const isValidFunction = input.prop<(v: string) => boolean>('isValid');
        return expect(isValidFunction(value));
    };

    changeInput = (label: string, value: string) => {
        const input = this.wrapper.find(`Input[label='${label}']`).first();
        const onChangeFn = input.prop<(v: string) => void>('onChange');
        onChangeFn(value);
    };

}