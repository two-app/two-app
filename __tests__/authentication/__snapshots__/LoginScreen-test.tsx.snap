// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should maintain snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text>
    Sign In
  </Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Sign Up
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
